### DISCUSSION TITLE
Need help creating a new Project and testing it locally

### DISCUSSION THREAD

**Discussion by caopengau** [aws/aws-pdk#708] [Q&A] [answered]
Following the [contributing guide here](https://aws.github.io/aws-pdk/contributing/index.html), it's a bit unclear what to do next after
`Once your class is created, simply instantiate it within .projenrc.ts and run pdk from the root in order for it to be synthesized.`
(now sure how to writing new project and testing it)

This is what I have attempted:
In the new generated `package/my-package` project, i have added `src/index.ts` with basic project construct like below (following projen contributing [example](https://projen.io/docs/concepts/projects/building-your-own))
```
export interface MyMicroserviceProjectOptions extends TypeScriptProjectOptions {
  readonly prMention?: string;
}

export class MyMicroserviceProject extends TypeScriptProject {
  constructor(options: MyMicroserviceProjectOptions) {
    super({
      ...options,
      pullRequestTemplate: false,
    });
  }
}
```

I couldn't get it build correctly, when i run `pdk build` at root or under `packages/pdk`, the execution fails with error below. How to fix this:
```
👾 package | monorepo.pnpm-link-bundled-transitive-deps packages/pdk
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[Error: ENOENT: no such file or directory, stat '/Users/pengcao/workspace/aws-pdk/packages/pdk/node_modules/@pnpm/dependency-path'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/Users/pengcao/workspace/aws-pdk/packages/pdk/node_modules/@pnpm/dependency-path'
}

Node.js v20.11.1
👾 Task "package" failed when executing "monorepo.pnpm-link-bundled-transitive-deps packages/pdk" (cwd: /Users/pengcao/workspace/aws-pdk/packages/pdk)
 ELIFECYCLE  Command failed with exit code 1.
/Users/pengcao/.nvm/versions/node/v20.11.1/lib/node_modules/@aws/pdk/node_modules/execa/index.js:212
		throw error;
		^

Error: Command failed with exit code 1: pnpm package
    at makeError (/Users/pengcao/.nvm/versions/node/v20.11.1/lib/node_modules/@aws/pdk/node_modules/execa/lib/error.js:60:11)
    at module.exports.sync (/Users/pengcao/.nvm/versions/node/v20.11.1/lib/node_modules/@aws/pdk/node_modules/execa/index.js:194:17)
    at module.exports.commandSync (/Users/pengcao/.nvm/versions/node/v20.11.1/lib/node_modules/@aws/pdk/node_modules/execa/index.js:235:15)
    at Object.<anonymous> (/Users/pengcao/.nvm/versions/node/v20.11.1/lib/node_modules/@aws/pdk/_scripts/exec-command.js:37:9)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
    at node:internal/main/run_main_module:28:49 {
  shortMessage: 'Command failed with exit code 1: pnpm package',
  command: 'pnpm package',
  escapedCommand: 'pnpm package',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: undefined,
  stderr: undefined,
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
```

Also, for next step, how do I link the locally built pdk and start testing MyMicroserviceProject in /my/other/project folder?

Thank you!

**Comment by cogwirrel** [comment, marked as answer]
Hey Cao,

Thanks for reaching out! Are you looking to contribute a new package to PDK? Or do you just intend to use PDK for a project of your own?

If `MyMicroserviceProject` is something you're not looking to contribute to PDK but rather something you'd like as part of your project, you can just define it in your `.projenrc.ts` for that project, or even in a separate file like `projenrc/my-microservice-project.ts` and consume it in your `.projenrc.ts`.

If you're really looking to contribute a new module to PDK, here are some thoughts below:

I've seen an issue like that before when building PDK and it was because I had some old `node_modules` - usually cleaning up the untracked files sorts this out (eg `git clean -fXd`), then reinstalling and building with `pnpm i && pnpm build`. Be careful if you want to keep your new PDK package though as the `git clean` command will delete uncommitted code.

For consuming local PDK in a test project, I'll often do something like this in my `.projenrc.ts`:

```ts
const monorepo = new MonorepoTsProject({
  packageManager: NodePackageManager.PNPM,
  devDeps: [
    "@aws/pdk@file:/path/to/pdk/packages/pdk/dist/js/package",
  ],
  name: "test",
  projenrcTs: true,
});

monorepo.package.addPackageResolutions("@aws/pdk@file:/path/to/pdk/packages/pdk/dist/js/package");
```

To iterate I just run a build in `packages/pdk`, and then extract the `tar.gz` file in `packages/pdk/dist/js`.

Hope that helps,
Jack

**Reply by caopengau** [reply]
Thanks for your reply and I will try your suggestions. I am looking for both contribute a new package and use PDK coming projects.

Do you think serverless framework template project will make a good candidate PR?

**Reply by cogwirrel** [reply]
No problem! It sounds like a serverless framework template project might actually be better suited in the [projen](https://github.com/projen/projen) codebase? Projen has a large collection of different project types and I can see serverless fitting better alongside those.

In order for a serverless package to fit in PDK it would need to integrate nicely with the other packages (for example Type Safe API), which would likely add too much complexity (eg generating snippets of serverless to create the api infrastructure rather than the type-safe CDK constructs). Hope that makes sense!

With a serverless framework project in projen, you'd still be able to nicely integrate it into a PDK monorepo etc :)
